src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/EETypeNode.cs

```diff
@@ -600,14 +629,67 @@ public override bool HasDynamicDependencies
                 // This node's EETypeOptionalFields node may change if this EEType implements interfaces
                 // that are used since the dispatch map table index is computed once we know the interface
                 // layout later on in compilation.
-                return _type.RuntimeInterfaces.Length > 0 && _constructed;
+
+                // Also, if any of the implemented interfaces have variance, variant dispatch might result in
+                // slots being necessary.
+                foreach (var intfType in _type.GetClosestMetadataType().RuntimeInterfaces)
+                {
+                    if (intfType.GetTypeDefinition().HasGenericVariance())
+                    {
+                        _variantInterfacesImplemented.Add(intfType);
+                    }
+                }
+
+                return _type.GetClosestMetadataType().RuntimeInterfaces.Length > 0 && _constructed;
             }
         }
 
         public override IEnumerable<CombinedDependencyListEntry> SearchDynamicDependencies(List<DependencyNodeCore<NodeFactory>> markedNodes, int firstNode, NodeFactory factory)
         {
             List<CombinedDependencyListEntry> dynamicNodes = new List<DependencyNodeCore<NodeFactory>.CombinedDependencyListEntry>();
             dynamicNodes.Add(new CombinedDependencyListEntry(factory.EETypeOptionalFields(_optionalFieldsBuilder), null, "EEType optional fields"));
+
+            // If the type implements any variant interfaces, we need to look for uses of interface methods that are compatible
+            // with the interface implemented by this type.
+            // We can avoid this work if we're not using a demand built VTable (producing a full type)
+            if (_variantInterfacesImplemented.Count > 0 && !factory.CompilationModuleGroup.ShouldProduceFullType(_type))
+            {
+                Debug.Assert(_constructed);
+
+                for (int i = firstNode; i < markedNodes.Count; i++)
+                {
+                    var node = markedNodes[i] as ReadyToRunHelperNode;
+                    if (node != null
+                        && (node.Id == ReadyToRunHelperId.InterfaceDispatch || node.Id == ReadyToRunHelperId.ResolveVirtualFunction))
+                    {
+                        MethodDesc targetMethod = (MethodDesc)node.Target;
+                        if (targetMethod.OwningType.IsInterface)
+                        {
+                            TypeDesc targetInterface = targetMethod.OwningType;
+                            // TODO: we might need to cache the result of "is this interface compatible with this type?"
+                            for (int j = 0; j < _variantInterfacesImplemented.Count; j++)
+                            {
+                                // TODO: this should really use CanCastTo/IsAssignable
+                                if (targetInterface.GetTypeDefinition() == _variantInterfacesImplemented[j].GetTypeDefinition())
```

Comment:
I'll make CanCastTo a separate pull request to keep things clean. I'll start on it tomorrow in parallel.